### PR TITLE
Fix yaml encoder for store accessor

### PR DESCRIPTION
**Pull request by kamipo**
Currently CI is broken due to 56a61e0c439d75244992f01468f09f85b5b08b78 and c4cb6862babd2665a65056e205c2a5fd17a5d99d.

This PR includes #25234 and to fix c4cb6862babd2665a65056e205c2a5fd17a5d99d.

### CODE REVIEW

**Comment by maclover7** [comment]
👍, this fixes Active Record's failure, but Active Model still needs to be sorted out 😬

**Comment by kamipo** [comment]
Passed tests!

**Comment by maclover7** [comment]
🎉

File: activerecord/lib/active_record/attribute_set/yaml_encoder.rb [hunk 1]
```diff
@@ -9,7 +9,7 @@ def initialize(default_types)
 
       def encode(attribute_set, coder)
         coder['concise_attributes'] = attribute_set.each_value.map do |attr|
-          if attr.type.equal?(default_types[attr.name])
+          if attr.type.equal?(default_types[attr.name]) && !attr.changed?
```
Comment (sgrif) [inline activerecord/lib/active_record/attribute_set/yaml_encoder.rb]: What does this fix? Why does it fix it? This change feels unrelated to the intent of the existing code. Can you give more context as to why this fixes the root problem and isn't a bandaid?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb [hunk 2]
```diff
@@ -53,7 +53,7 @@ def custom_primary_key?
 
       def defined_for?(options_or_to_table = {})
         if options_or_to_table.is_a?(Hash)
-          options_or_to_table.all? {|key, value| options[key].to_s == value.to_s }
+          options_or_to_table.all? {|key, value| send(key).to_s == value.to_s }
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_definitions.rb]: What does this fix? Why does it fix it? This change feels unrelated to the intent of the existing code. Can you give more context as to why this fixes the root problem and isn't a bandaid?

File: activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb [hunk 3]
```diff
@@ -855,6 +855,7 @@ def remove_reference(table_name, ref_name, foreign_key: false, polymorphic: fals
           else
             foreign_key_options = { to_table: reference_name }
           end
+          foreign_key_options[:column] ||= "#{ref_name}_id"
```
Comment (sgrif) [inline activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb]: Can you give context as to what this is doing and why it's needed? The code prior to the referenced commit didn't include this either, why was it working before?

**Comment by sgrif** [comment]
Normally I'd wait for response to feedback, but since the build was failing I've gone ahead and updated. Closing in favor of 1b8a7b8. You've been credited on that commit.
